core/src/indexer/native_transfer.rs

```diff
@@ -137,38 +140,182 @@ pub async fn native_transfer_block_fetch(
             }
             Ok(None) => {}
             Err(e) => {
-                error!("Error fetching '{}' block traces: {}", network, e.to_string());
+                error!("Error fetching '{}' blocks: {}", network, e.to_string());
                 sleep(Duration::from_secs(1)).await;
             }
         }
     }
 }
 
-async fn provider_call(
+pub async fn native_transfer_block_processor(
+    network_name: String,
+    provider: Arc<JsonRpcCachedProvider>,
+    config: Arc<TraceProcessingConfig>,
+    mut block_rx: mpsc::Receiver<U64>,
+) -> Result<(), ProcessEventError> {
+    let is_rcp_batchable = config.method == TraceProcessingMethod::EthGetBlockByNumber;
+
+    // Set the concurrency used to make requests based on the method.
+    //
+    // Currently, `eth_getBlockByNumber` is a single JSON-RPC batch, and others are individual
+    // network calls so can be treated differently.
+    let (initial_concurrent_requests, limit_concurrent_requests) =
+        if is_rcp_batchable { (100, RPC_CHUNK_SIZE) } else { (5, 100) };
+
+    let mut concurrent_requests: usize = initial_concurrent_requests;
+    let mut buffer: Vec<U64> = Vec::with_capacity(limit_concurrent_requests);
+
+    loop {
+        let recv = block_rx.recv_many(&mut buffer, concurrent_requests).await;
+
+        if recv == 0 {
+            sleep(Duration::from_secs(1)).await;
+            continue;
+        }
+
+        let processed_block = native_transfer_block_consumer(
+            provider.clone(),
+            &buffer[..recv],
+            &network_name,
+            &config,
+        )
+        .await;
+
+        // If this has an error, we need to not and reconsume the blocks. We don't have
+        // to worry about double-publish because the failure point is on the provider
+        // call itself, which is before publish.
+        if let Err(e) = processed_block {
+            // On error, drop the block query range. We want a slow increase in concurrency and a
+            // relatively aggressive backoff.
+            concurrent_requests = cmp::max(1, (concurrent_requests as f64 * 0.8) as usize);
+
+            let is_rate_limit_error = matches!(&e, ProcessEventError::ProviderCallError(
+                ProviderError::RequestFailed(
+                    RpcError::Transport(
+                        TransportErrorKind::HttpError(http_err)
+                    )
+                )) if http_err.status == 429
+            );
+
+            if is_rate_limit_error {
+                error!(
```

Comment:
is this an error feels just like a warning..

Reply:
Yeah, warn is better.